stdlib/public/runtime/ImageInspectionELF.cpp

```diff
@@ -35,67 +39,116 @@ static const char ProtocolConformancesSymbol[] =
+    addImageProtocolConformanceBlockCallback(block.data, block.size);
+  } else {
+    protocolConformanceDSOAddrs->push_back(addr);
+  }
 }
 
 void swift::initializeProtocolConformanceLookup() {
-  // Search the loaded dls. This only searches the already
-  // loaded ones.
-  // FIXME: Find a way to have this continue to happen for dlopen-ed images.
-  // rdar://problem/19045112
-  InspectArgs ProtocolConformanceArgs = {
-    ProtocolConformancesSymbol,
-    addImageProtocolConformanceBlockCallback
-  };
-  dl_iterate_phdr(iteratePHDRCallback, &ProtocolConformanceArgs);
+  assert(protocolConformanceLock != nullptr);
+  ScopedLock guard(*protocolConformanceLock);
+  if (protocolConformanceDSOAddrs) {
+    for (auto addr: *protocolConformanceDSOAddrs) {
+      auto block = getSectionInfo(addr, ProtocolConformancesSymbol);
+      addImageProtocolConformanceBlockCallback(block.data, block.size);
+    }
+    delete protocolConformanceDSOAddrs;
+    protocolConformanceDSOAddrs = nullptr;
+  }
+}
+
+static void lazyInitTypeMetadataRecord(const void *addr) {
+  ScopedLock guard(*typeMetadataLock);
+  if (typeMetadataRecordDSOAddrs == nullptr) {
+    auto block = getSectionInfo(addr, TypeMetadataRecordsSymbol);
+    addImageTypeMetadataRecordBlockCallback(block.data, block.size);
+  } else {
+    typeMetadataRecordDSOAddrs->push_back(addr);
+  }
 }
 
 void swift::initializeTypeMetadataRecordLookup() {
-  InspectArgs TypeMetadataRecordArgs = {
-    TypeMetadataRecordsSymbol,
-    addImageTypeMetadataRecordBlockCallback
-  };
-  dl_iterate_phdr(iteratePHDRCallback, &TypeMetadataRecordArgs);
+  assert(typeMetadataLock != nullptr);
+  ScopedLock guard(*typeMetadataLock);
+  if (typeMetadataRecordDSOAddrs) {
+    for (auto addr: *typeMetadataRecordDSOAddrs) {
+      auto block = getSectionInfo(addr, TypeMetadataRecordsSymbol);
+      addImageTypeMetadataRecordBlockCallback(block.data, block.size);
+    }
+    delete typeMetadataRecordDSOAddrs;
+    typeMetadataRecordDSOAddrs = nullptr;
+  }
+}
+
+// As ELF images are loaded, ImageInspectionInit.cpp will call addNewDSOImage()
+// with an address in the DSO that can later be used to dlopen() the image.
+// The address is just saved until initialise*Lookup() is called and then
+// the section data can be extracted from each image as needed
+
+SWIFT_RUNTIME_EXPORT
+void swift::addNewDSOImage(const void *addr) {
+  static OnceToken_t onceToken;
+  SWIFT_ONCE_F(onceToken, [](void *context) {
+      protocolConformanceLock = new Mutex();
+      protocolConformanceDSOAddrs = new vector<const void *>();
+
+      typeMetadataLock = new Mutex();
+      typeMetadataRecordDSOAddrs = new vector<const void *>();
+
+      // Add the main executable to the lists
+      protocolConformanceDSOAddrs->push_back(nullptr);
+      typeMetadataRecordDSOAddrs->push_back(nullptr);
+    }, nullptr);
+
+  lazyInitProtocolConformance(addr);
+  lazyInitTypeMetadataRecord(addr);
```

Comment:
Code organization looks much better, thanks! Keep in mind though that `__attribute__((constructor))`s fire not only at `dlopen` time, but at process load time, so all this laziness will be wasted when `addNewDSOImage` gets invoked early at process start. We need to keep track of whether each cache has been initialized to avoid doing any work if it hasn't. Something like this:

```
static std::atomic<bool> didInitializeProtocolConformanceLookup;
void swift::initializeProtocolConformanceLookup() {
  // Set `didInitializeProtocolConformanceLookup` so that loading
  // future Swift images will register with the lookup cache.
  didInitializeProtocolConformanceLookup.store(true, std::memory_order_release);
  // Search the loaded dls...
  ...
}

extern "C"
void swift::swift_addNewDSOImage(const void *addr) {
  if (didInitializeProtocolConformanceLookup.load(std::memory_order_acquire)) {
    addImageProtocolConformanceBlockCallback(...get section from addr...);
  }
}
```

and a similar pattern for the type metadata lookup cache.